### ISSUE TITLE
RDSE max number of buckets is hardcoded and confusing

### ISSUE THREAD

**Issue by chetan51** [numenta/nupic#1951]
This behavior of the RDSE confused me:

	In [2]: s = RandomDistributedScalarEncoder(0.1, w=21, n=1024)

	In [3]: s.encode(0)
	Out[3]: array([0, 0, 0, ..., 0, 0, 0], dtype=uint8)

	In [4]: s.encode(0).nonzero()[0]
	Out[4]:
	array([ 31,  70,  76,  96, 101, 280, 357, 370, 404, 425, 444, 447, 525,
	       587, 618, 621, 653, 677, 715, 734, 749])

	In [5]: s.encode(1).nonzero()[0]
	Out[5]:
	array([ 70,  76, 101, 136, 235, 280, 317, 333, 370, 375, 425, 458, 580,
	       621, 653, 677, 715, 734, 884, 945, 955])

	In [6]: s.encode(2).nonzero()[0]
	Out[6]:
	array([136, 235, 295, 317, 319, 333, 348, 375, 425, 458, 512, 546, 571,
	       580, 831, 884, 917, 929, 945, 955, 994])

	In [7]: s.encode(99).nonzero()[0]
	Out[7]:
	array([ 16,  38,  60,  72, 107, 130, 153, 171, 200, 216, 228, 364, 449,
	       481, 497, 506, 615, 640, 712, 848, 938])

	In [8]: s.encode(98).nonzero()[0]
	Out[8]:
	array([ 16,  38,  60,  72, 107, 130, 153, 171, 200, 216, 228, 364, 449,
	       481, 497, 506, 615, 640, 712, 848, 938])

	In [9]: s.encode(97).nonzero()[0]
	Out[9]:
	array([ 16,  38,  60,  72, 107, 130, 153, 171, 200, 216, 228, 364, 449,
	       481, 497, 506, 615, 640, 712, 848, 938])

Note that 99, 98, and 97 all have identical encodings, though the resolution is 0.1.

@subutai noted that this is because there is a hard coded max number of buckets = 1000. Anything beyond that gives the same representation.

This is hidden and therefore confusing.

**Comment by subutai** [comment]
The rationale for this is given at the bottom:
```
  def _initializeBucketMap(self, maxBuckets, offset):
    """
    Initialize the bucket map assuming the given number of maxBuckets.
    """
    # The first bucket index will be _maxBuckets / 2 and bucket indices will be
    # allowed to grow lower or higher as long as they don't become negative.
    # _maxBuckets is required because the current CLA Classifier assumes bucket
    # indices must be non-negative. This normally does not need to be changed
    # but if altered, should be set to an even number.
```
To make it less confusing we could make maxBuckets be a constructor parameter and explicitly document it.

**Comment by cogmission** [comment]
Not sure if this is what you guys were thinking, but I would say to pre compute the smallest/largest encoded value (before clipping begins) then check it in the encode() method and out out a log statement or exception depending on the severity you want to enforce - thatc at the user is never surprised by code that has a "personality" :-)

**Comment by cogmission** [comment]
*... and output a log statement

*- that way the user is never surprised by code that has a "personality"

Sorry mobile device

**Comment by breznak** [comment]
you can't do that for streaming data. If you can and do know the min/max in
your dataset, there's no point in "adaptive" encoder, better use ScalarEnc.

**Comment by cogmission** [comment]
As I understood it there's a max implied by the number of buckets the encoder was encoding? so at a resolution of .1 the maximum number that you could have with 1000 buckets is 99?

**Comment by breznak** [comment]
> As I understood it there's a max implied by the number of buckets the
encoder was encoding? so at a resolution of .1 the maximum number that you
could have with 1000 buckets is 99?

> then check it in the encode() method and out out a log statement or
exception depending on the severity you want to enforce -

My bad, now I understand it, sorry :)

But then..this is quite severe problem...

**Comment by chetan51** [comment]
I believe the max would be relative to the first number that was encoded.

**Comment by cogmission** [comment]
So if the first number you encounter is the "highest" number, then all the
rest of the numbers would be capped to originalN - 99 correct? (given the
same 0.1 resolution) - so is it more accurate for me to say that the "span"
of numbers is always "resolution * max_buckets"?

Either case, the decision is how to treat it so that the user has no
"surprises" right? So I like your original concern and I suggest that it is
handled as soon as possible so that the user can reconfigure things if
necessary? This is probably obvious though... sorry...



On Wed, Mar 25, 2015 at 4:40 PM, Chetan Surpur <notifications@github.com>
wrote:

> I believe the max would be relative to the first number that was encoded.
> On Mar 25, 2015 2:34 PM, "David Ray" <notifications@github.com> wrote:
>
> > As I understood it there's a max implied by the number of buckets the
> > encoder was encoding? so at a resolution of .1 the maximum number that
> you
> > could have with 1000 buckets is 99?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/numenta/nupic/issues/1951#issuecomment-86225424>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/1951#issuecomment-86227079>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by breznak** [comment]
I like the ideas of adding `nBuckets` as an opt. param, + David's issue a
warning when cropped (maybe even with init param setting
`onOverflow="warn/err"` where the user could decide to throw an error
instead.)

Under the new light, is `AdaptiveScalar` superior to RDSE? Can the
functionality of both be combined?

**Comment by subutai** [comment]
Yeah if we throw an exception it should probably be optional. We can also make the default number of buckets quite a bit larger I think (say 5000), which would mitigate the issue. I was worried about performance penalties but in practice I think they are probably minor.

>is AdaptiveScalar superior to RDSE?

The issue with AdaptiveScalar is that the incoming SDR changes with experience. That can be ok in some situations if the changes are very slow. In practice it often causes problems.  I think this could be a good research area.
